src/Sylius/Bundle/LocaleBundle/Listener/RequestLocaleSetter.php

```diff
@@ -0,0 +1,52 @@
+<?php
+
+namespace Sylius\Bundle\LocaleBundle\Listener;
+
+use Sylius\Component\Locale\Context\LocaleContextInterface;
+use Sylius\Component\Locale\Context\LocaleNotFoundException;
+use Sylius\Component\Locale\Provider\LocaleProviderInterface;
+use Symfony\Component\HttpKernel\Event\GetResponseEvent;
+
+/**
+ * @author Kamil Kokot <kamil.kokot@lakion.com>
+ */
+final class RequestLocaleSetter
+{
+    /**
+     * @var LocaleContextInterface
+     */
+    private $localeContext;
+
+    /**
+     * @var LocaleProviderInterface
+     */
+    private $localeProvider;
+
+    /**
+     * @param LocaleContextInterface $localeContext
+     * @param LocaleProviderInterface $localeProvider
+     */
+    public function __construct(
+        LocaleContextInterface $localeContext,
+        LocaleProviderInterface $localeProvider
+    ) {
+        $this->localeContext = $localeContext;
+        $this->localeProvider = $localeProvider;
+    }
+
+    /**
+     * @param GetResponseEvent $event
+     */
+    public function onKernelRequest(GetResponseEvent $event)
+    {
+        $request = $event->getRequest();
+
+        try {
+            $request->setLocale($this->localeContext->getLocaleCode());
+        } catch (LocaleNotFoundException $exception) {}
+
+        try {
+            $request->setDefaultLocale($this->localeProvider->getDefaultLocaleCode());
+        } catch (LocaleNotFoundException $exception) {}
```

Comment:
Had some reasoning about it, but can't recall it. I'll take a look & think how to manage this in #5601.